### PR TITLE
Add autoindent support

### PR DESCRIPTION
**Pull request by rymndhng**
Addresses #28 

I took the suggestion from the comments and worked this from https://github.com/skuro/plantuml-mode/issues/28#issuecomment-276821796.

- simplified implementation (I think)
- I tweaked the regex patterns to support additional syntax such as `alt` and `note of`. 
- I added tests by taking ideas from clojure-mode's indent test

NOTE: I'm a emacs-lisp library noob and need guidance on how to clean this up.

### CODE REVIEW

File: plantuml-mode.el [hunk 1]
```diff
@@ -373,6 +373,40 @@ Uses prefix (as PREFIX) to choose where to display it:
                 (all-completions meat plantuml-kwdList)))
              (message "Making completion list...%s" "done")))))
 
+
+;; indentation
+(defvar plantuml-indent-regexp-start "^[ \t]*\\(\\(?:.*\\)?\s*\\(?:[<>.*a-z-|]+\\)?\s*\\(?:\\[[a-zA-Z]+\\]\\)?\s+if\\|alt\\|else\\|note\s+over\\|note\sas\s.*\\|note\s+\\(\\(?:\\(?:buttom\\|left\\|right\\|top\\)\\)\\)\\(?:\s+of\\)?\\|\\(?:class\\|enum\\|package\\)\s+.*{\\)")
+
+(defvar plantuml-indent-regexp-end "^[ \t]*\\(endif\\|else\\|end\\|end\s+note\\|.*}\\)")
```
Comment (rymndhng) [inline plantuml-mode.el]: I feel like there's a better way of doing these regexes, but I'm not sure how 😬 . Also it feels like this should be unified with the other regexes defined above, but again I'm not quite sure how to go about it.

File: plantuml-mode.el [hunk 2]
```diff
@@ -373,6 +373,40 @@ Uses prefix (as PREFIX) to choose where to display it:
...
+
+(defun plantuml-current-block-indentation ()
+  (save-excursion
+    (let ((relative-depth 0))
+      (while (>= relative-depth 0)
+        ;; (message (concat "got: " (number-to-string relative-depth)))
```
Comment (skuro) [inline plantuml-mode.el]: remove comment please

File: plantuml-mode.el [hunk 3]
```diff
@@ -373,6 +373,40 @@ Uses prefix (as PREFIX) to choose where to display it:
...
+        (if (looking-at plantuml-indent-regexp-start)
+            (setq relative-depth (1- relative-depth))))
+
+      (if (= -2 relative-depth)
+          0
+        (+ 2 (current-indentation))))))
+
+(defun plantuml-indent-line ()
+  (interactive)
+  (beginning-of-line)
```
Comment (skuro) [inline plantuml-mode.el]: better wrap the body of `plantuml-indent-line` into `save-excursion`

File: plantuml-mode.el [hunk 4]
```diff
@@ -373,6 +373,40 @@ Uses prefix (as PREFIX) to choose where to display it:
...
+        (if (bobp)
+            (setq relative-depth -2))   ;end fast
+        (if (looking-at plantuml-indent-regexp-end)
+            (setq relative-depth (1+ relative-depth)))
+        (if (looking-at plantuml-indent-regexp-start)
+            (setq relative-depth (1- relative-depth))))
+
+      (if (= -2 relative-depth)
+          0
+        (+ 2 (current-indentation))))))
```
Comment (skuro) [inline plantuml-mode.el]: use `tab-width` instead of `2`

File: plantuml-mode.el [hunk 5]
```diff
@@ -373,6 +373,40 @@ Uses prefix (as PREFIX) to choose where to display it:
...
+
+(defun plantuml-indent-line ()
+  (interactive)
+  (beginning-of-line)
+  (if (bobp)
+      (indent-line-to 0)
+    (let ((offset (plantuml-current-block-indentation)))
+      (when (looking-at plantuml-indent-regexp-end)
+        (setq offset (max (- offset 2) 0)))
```
Comment (skuro) [inline plantuml-mode.el]: use `tab-width` instead of `2`

File: plantuml-mode.el (hunk 1 above)
Comment (skuro) [reply plantuml-mode.el]: Indeed, I would move these `defvar`s just after the definition of `plantuml-preprocessors-regexp`

File: test/plantuml-indentation-test.el [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+point.
+AFTER is the expected buffer string after indenting, where a pipe (|)
+represents the expected position of point.
+VAR-BINDINGS is an optional let-bindings list.  It can be used to set the
+values of customisable variables."
+  (declare (indent 1))
+  (let ((fname (intern (format "indentation/%s" description))))
+    `(ert-deftest ,fname ()
+       (let* ((after ,after)
+              (clojure-indent-style :always-align)
```
Comment (skuro) [inline test/plantuml-indentation-test.el]: this line can go

**Comment by skuro** [comment]
@rymndhng thanks to you! I'm actually still busy fixing the tests, for some reasons they won't pick up the custom `plantuml-jar-path` setting and always fail to initialize the mode. did the tests work for you?

**Comment by rymndhng** [comment]
@skuro Hmm, how are you running the tests? Admittedly I was doing it from inside emacs using `ert-run-tests-interactively`.

When i was developing, I ran into some issues attempting to redefine a `defvar` (which behaves like clojure's `defonce`). Could that be messing with the test runner?

**Comment by skuro** [comment]
Same as CI, I'm running them with `make`, which pretty much runs `cask install && cask exec ert-runner`

**Comment by skuro** [comment]
Mystery [solved](https://github.com/skuro/plantuml-mode/commit/6c0455b5e2e105a2ac405549883bf979c3487b90)! I ended up rewriting good parts of the implementation, but that was good elisp training ;-)

**Pull request merged**